extras/templates/model.template.jinja2

```diff
@@ -0,0 +1,345 @@
+	column values to do fast bit-wise filtering.
+
+	A S2 geo index in created for lat, lon values.
+
+	Unique values are stored as-is.
+
+	The generated codes leaves room to create custom
+	index functions yourself to create an API with an
+	< 1 ms response time for your specific needs.
+
+	This codebase solves: I need to have an API on this
+	tabular dataset fast!
+*/
+
+package main
+
+import (
+	"log"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"errors"
+
+	"github.com/Workiva/go-datastructures/bitarray"
+)
+
+type registerGroupByFunc map[string]func(*Item) string
+type registerGettersMap map[string]func(*Item) string
+type registerReduce map[string]func(Items) map[string]string
+
+type registerBitArray map[string]func(s string) (bitarray.BitArray, error)
+
+type fieldIdxMap map[string]uint16
+type fieldMapIdx map[uint16]string
+type fieldItemsMap map[uint16]bitarray.BitArray
+
+// Column maps.
+// Store for each non distinct/repeated column
+// unit16 -> string map and
+// string -> unit16 map
+// track count of distinct values
+
+{{shrinkVars}}
+
+/*
+var {columnname}Tracker uint16
+var {columnname}IdxMap fieldIdxMap
+var {columnname} fieldMapIdx
+var {columnname}Items fieldItemmap
+*/
+
+// item map lock
+var lock = sync.RWMutex{}
+
+// bitArray Lock
+var balock = sync.RWMutex{}
+
+
+func init() {
+
+	{{initRepeatColumns}}
+
+	/*
+	labelscoredefinitiefTracker = 0
+	labelscoredefinitiefIdxMap = make(fieldIdxMap)
+	labelscoredefinitief = make(fieldMapIdx)
+	*/
+}
+
+{{itemStructs}}
+
+
+type ItemIn struct {
+
+{{columnsItemIn}}
+
+}
+
+type ItemOut struct {
+
+{{columnsItemOut}}
+
+}
+
+
+type Item struct {
+
+	Label                int  // internal index in ITEMS
+{{columnsItem}}
+
+}
+
+func (i Item) MarshalJSON() ([]byte, error) {
+	return json.Marshal(i.Serialize())
+}
+
+// Shrink create smaller Item using uint16
+func (i ItemIn) Shrink(label int) Item {
+
+	lock.Lock()
+	defer lock.Unlock()
+
+{{shrinkItems}}
+
+	return Item{
+
+		label,
+
+{{shrinkItemFields}}
+
+	}
+}
+
+// Store selected columns in seperate map[columnvalue]bitarray
+// for gast item lookup
+func (i Item) StoreBitArrayColumns() {
+
+	balock.Lock()
+	defer balock.Unlock()
+
+	lock.RLock()
+	defer lock.RUnlock()
+
+        {% if bitArrayStores is defined %}
+	var ba bitarray.BitArray
+	var ok bool
+
+	{{ bitArrayStores }}
+
+	{% endif %}
+
+	/*
+	// Column Buurtcode has byte arrays for
+	ba, ok = BuurtcodeItems[i.Buurtcode]
+	if !ok {
+		ba = bitarray.NewSparseBitArray()
+		BuurtcodeItems[i.Buurtcode] = ba
+	}
+	ba.SetBit(uint64(i.Label))
+	*/
+
+}
+
+func (i Item) Serialize() ItemOut {
+
+	lock.RLock()
+	defer lock.RUnlock()
+
+	return ItemOut{
+
+{{expandItemFields}}
+
+	}
+}
+
+func (i ItemIn) Columns() []string {
+	return []string{
+
+{{inColumns}}
+
+	}
+}
+
+func (i ItemOut) Columns() []string {
+	return []string{
+
+{{outColumns}}
+
+	}
+}
+
+
+func (i Item) Row() []string {
+
+	lock.RLock()
+	defer lock.RUnlock()
```

Comment:
Since we use a translation from internal Items to ItemsOut while still reading or creating a race condition can occur then doing a translation from Item to ItemOut. Thats why this read lock is needed. It doesn't cause a notible delay IMHO. only used to output 'some' rows in the end for viewing. It is not executed millions of times.